File: backend/app/calculations.py
from .models import Scenario


def simulate(scenario: Scenario) -> dict:
    contribution_per_unit = (
        scenario.price -
        scenario.variable_cost
    )

    revenue = (
        scenario.units *
        scenario.price
    )

    variable_cost_total = (
        scenario.units *
        scenario.variable_cost
    )

    contribution = (
        scenario.units *
        contribution_per_unit
    )

    contribution_margin_percent = (
        contribution_per_unit /
        scenario.price *
        100
        if scenario.price > 0
        else 0
    )

    if contribution_per_unit > 0:
        break_even_units = (
            scenario.fixed_cost /
            contribution_per_unit
        )
    else:
        break_even_units = None

    if break_even_units is not None:
        break_even_revenue = (
            break_even_units *
            scenario.price
        )
    else:
        break_even_revenue = None

    operating_profit = (
        contribution -
        scenario.fixed_cost
    )

    if break_even_units is not None:
        safety_units = (
            scenario.units -
            break_even_units
        )
    else:
        safety_units = None

    if (
        break_even_units is not None
        and scenario.units > 0
    ):
        safety_margin_percent = (
            (
                scenario.units -
                break_even_units
            )
            /
            scenario.units
            *
            100
        )
    else:
        safety_margin_percent = None

    ltv = (
        scenario.monthly_revenue *
        scenario.gross_margin_percent *
        scenario.retention_months
    )

    if scenario.cac > 0:
        ltv_cac_ratio = (
            ltv /
            scenario.cac
        )
    else:
        ltv_cac_ratio = None

    return {
        "data_status": "synthetic",

        "revenue": revenue,

        "variable_cost_total":
            variable_cost_total,

        "contribution":
            contribution,

        "contribution_per_unit":
            contribution_per_unit,

        "contribution_margin_percent":
            contribution_margin_percent,

        "fixed_cost":
            scenario.fixed_cost,

        "operating_profit":
            operating_profit,

        "break_even_units":
            break_even_units,

        "break_even_revenue":
            break_even_revenue,

        "safety_units":
            safety_units,

        "safety_margin_percent":
            safety_margin_percent,

        "ltv":
            ltv,

        "cac":
            scenario.cac,

        "ltv_cac_ratio":
            ltv_cac_ratio
    }

File: frontend/lib/calculations.ts
import {
  Scenario,
  SimulationResult
} from "./types";

export function calculateSimulation(
  scenario: Scenario
): SimulationResult {
  const price = Math.max(0, scenario.price);

  const variableCost = Math.max(
    0,
    scenario.variableCost
  );

  const fixedCost = Math.max(
    0,
    scenario.fixedCost
  );

  const units = Math.max(
    0,
    scenario.units
  );

  const cac = Math.max(
    0,
    scenario.cac
  );

  const monthlyRevenue = Math.max(
    0,
    scenario.monthlyRevenue
  );

  const retentionMonths = Math.max(
    0,
    scenario.retentionMonths
  );

  const grossMarginPercent = Math.max(
    0,
    Math.min(1, scenario.grossMarginPercent)
  );

  const contributionPerUnit =
    price - variableCost;

  const revenue =
    units * price;

  const variableCostTotal =
    units * variableCost;

  const contribution =
    units * contributionPerUnit;

  const contributionMarginPercent =
    price > 0
      ? (contributionPerUnit / price) * 100
      : 0;

  const breakEvenUnits =
    contributionPerUnit > 0
      ? fixedCost / contributionPerUnit
      : null;

  const breakEvenRevenue =
    breakEvenUnits !== null
      ? breakEvenUnits * price
      : null;

  const operatingProfit =
    contribution - fixedCost;

  const safetyUnits =
    breakEvenUnits !== null
      ? units - breakEvenUnits
      : null;

  const safetyMarginPercent =
    breakEvenUnits !== null && units > 0
      ? ((units - breakEvenUnits) / units) * 100
      : null;

  const ltv =
    monthlyRevenue *
    grossMarginPercent *
    retentionMonths;

  const ltvCacRatio =
    cac > 0
      ? ltv / cac
      : null;

  return {
    revenue,

    variableCostTotal,

    contribution,

    contributionPerUnit,

    contributionMarginPercent,

    fixedCost,

    operatingProfit,

    breakEvenUnits,

    breakEvenRevenue,

    safetyUnits,

    safetyMarginPercent,

    ltv,

    cac,

    ltvCacRatio
  };
}

export type SensitivityResult = {
  label: string;

  key: keyof Scenario;

  impact: number;

  positiveDirection: boolean;
};

export function calculateSensitivity(
  scenario: Scenario
): SensitivityResult[] {
  const drivers: Array<{
    label: string;
    key: keyof Scenario;
  }> = [
    {
      label: "Price",
      key: "price"
    },
    {
      label: "Variable cost",
      key: "variableCost"
    },
    {
      label: "Fixed cost",
      key: "fixedCost"
    },
    {
      label: "CAC",
      key: "cac"
    },
    {
      label: "Retention",
      key: "retentionMonths"
    }
  ];

  return drivers
    .map((driver) => {
      const baseValue =
        Number(scenario[driver.key]);

      const increaseScenario = {
        ...scenario,
        [driver.key]:
          baseValue * 1.1
      };

      const decreaseScenario = {
        ...scenario,
        [driver.key]:
          baseValue * 0.9
      };

      const increaseProfit =
        calculateSimulation(
          increaseScenario
        ).operatingProfit;

      const decreaseProfit =
        calculateSimulation(
          decreaseScenario
        ).operatingProfit;

      return {
        label: driver.label,

        key: driver.key,

        impact:
          Math.abs(
            increaseProfit -
              decreaseProfit
          ) / 2,

        positiveDirection:
          increaseProfit >=
          decreaseProfit
      };
    })
    .sort(
      (a, b) =>
        b.impact - a.impact
    );
}